src/realm/sync/noinst/client_reset.cpp

```diff
@@ -426,124 +446,231 @@ void remove_pending_client_resets(Transaction& wt)
+                                      PendingReset::Action action, const std::optional<Status>& error,
+                                      util::Logger& logger)
 {
     if (auto previous_reset = has_pending_reset(wt)) {
-        logger.info(util::LogCategory::reset, "A previous reset was detected of type: '%1' at: %2",
-                    previous_reset->type, previous_reset->time);
-        switch (previous_reset->type) {
-            case ClientResyncMode::Manual:
-                REALM_UNREACHABLE();
-            case ClientResyncMode::DiscardLocal:
-                throw ClientResetFailed(util::format("A previous '%1' mode reset from %2 did not succeed, "
-                                                     "giving up on '%3' mode to prevent a cycle",
-                                                     previous_reset->type, previous_reset->time, mode));
-            case ClientResyncMode::Recover:
-                switch (mode) {
-                    case ClientResyncMode::Recover:
-                        throw ClientResetFailed(util::format("A previous '%1' mode reset from %2 did not succeed, "
-                                                             "giving up on '%3' mode to prevent a cycle",
-                                                             previous_reset->type, previous_reset->time, mode));
-                    case ClientResyncMode::RecoverOrDiscard:
-                        mode = ClientResyncMode::DiscardLocal;
-                        logger.info(util::LogCategory::reset,
-                                    "A previous '%1' mode reset from %2 downgrades this mode ('%3') to DiscardLocal",
-                                    previous_reset->type, previous_reset->time, mode);
-                        remove_pending_client_resets(wt);
-                        break;
-                    case ClientResyncMode::DiscardLocal:
-                        remove_pending_client_resets(wt);
-                        // previous mode Recover and this mode is Discard, this is not a cycle yet
-                        break;
-                    case ClientResyncMode::Manual:
-                        REALM_UNREACHABLE();
-                }
-                break;
-            case ClientResyncMode::RecoverOrDiscard:
-                throw ClientResetFailed(util::format("Unexpected previous '%1' mode reset from %2 did not "
-                                                     "succeed, giving up on '%3' mode to prevent a cycle",
-                                                     previous_reset->type, previous_reset->time, mode));
+        logger.info(util::LogCategory::reset, "Found a previous %1", previous_reset->to_string());
+        if (previous_reset->error) {
+            logger.info(util::LogCategory::reset, "Originating client reset error: %1", *previous_reset->error);
+        }
+        if (action != previous_reset->action) {
```

Comment:
There's a functional change here that should be noted in the changelog. Is this a known/tracked issue you are fixing?

Reply:
Thanks for catching this - reverted.
I didn't mean to carry it over for this PR. This is intended for the PR #7542 that will address the bug that is actually being caught by the "Test client migration and rollback with recovery" test.

Reply:
Also reverted the `ClientReset_DifferentAction`, since it tested this functionality.